### DISCUSSION TITLE
[Chapter 07] 테스트클래스 빈 주입방식(416p)

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [cowaine/spring-boot-for-msa#68] [Chapter 07. Spring AOP, Testing, and Auto Confiugration] [answered]
## 테스트클래스 빈 주입방식(416p)
- 테스트클래스가 아닌 기존 클래스의 생성자주입방식에선 생성자에 `@Autowired` 가 붙어있진 않았는데, 
- 테스트클래스에선 빈주입을 생성자주입으로 할땐 `@Autowired`가 붙어있어야 한다고 합니다. 
- 왜 그럴까요?

```java
@SpringBootTest
class HotelDisplayServiceTest {

     private final HotelDisplayService hotelDisplayService;
     private final ApplicationContext applicationContext;
     
     @Autowired // 생성자주입방식으로 하려면 생성자에 Autowired필수
     public HotelDisplayServiceTest(HotelDisplayService hotelDisplayService,
                                    ApplicationContext applicationContext) {
          this.hotelDisplayService = hotelDisplayService;
          this.applicationContext = applicationContext;
     }
```

## TestPropertySource 설정 (p421)

```java
@SpringBootTest
//@Import (value={TestConfig.class}
@ContextConfiguration(classes = TestConfig.class) // Import 나 이거 사용
@TestPropertySource(locations = "classpath:application-test.yaml") // <<< 
class HotelRoomDisplayServiceTest01 {

```

```txt
spring:
  main:
    allow-bean-definition-overriding: true
```
- 위 처럼 `@TestPropertySource(locations` 설정을 해주고, 아래처럼 overriding true 를 했는데도 빈 재정의가 되지않아 테스트 실행이 제대로 되지 않았습니다. 
- 이후 `test/resources/application-test.yaml`외에,  `main/resources/application.yaml`의 false 를 true 로 바꾼 후에야 제대로 실행이 되었습니다.
- 제 생각엔 test관련 빈들에 대한 설정만 바꾸면 된다고 생각했었는데, main 폴더의 `application.yaml` 까지 바꿔줘야하는게 맞나요?

**Comment by Irisation23** [comment, marked as answer]
테스트클래스 빈 주입방식(416p) 에 대한 질문의 답을 드리고자 합니다.

먼저 테스트 클래스에서의 생성자 빈 주입 에러와 프로덕션 코드에서의 빈 주입 에러에 대해 비교 해 봅시다.

1. 테스트 클래스 생성자 빈 주입 에러

<img width="911" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/53285909/aedc8614-05ef-4b39-99d4-9ef02aa14adc">

- `NoParameterResolver` 에러가 발생함

2. 프로덕션 코드에서의 빈 주입 에러

<img width="1678" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/53285909/3f20e6b4-cd08-463a-9181-e6fff36637b4">

- `NoSuchBeanException` 에러가 발생함

자 해당 에러를 확인 해 보면 차이점이 있습니다.
테스트 코드에서 발생하는 에러는 `NoParameterResolver`, 프로덕션 코드에서의 빈 주입 에러는 `NoSuchBeanException`입니다.

이는 스프링 빈 주입의 방식이 프로덕션 코드와 테스트 코드간에 다른 방법을 취하고 있기 때문입니다.

생성자 매개변수의 경우 프로덕션 코드라면 **Spring IoC** 컨테이너가 이를 해결합니다.
스프링 프레임워크의 **ApplicationContext**는 등록할 빈들을 찾아서 저장했다가 적절한 시점에 빈을 생성자에 주입합니다.

그러나, 테스트 코드는 조금 다릅니다.

테스트 프레임워크에서의 생성자 매개변수 관리는 스프링 컨테이너가 아닌 `Jupiter`가 담당합니다.
따라서 `@Autowired`를 명시적으로 선언해야 `Jupiter`가 Spring Container에게 빈 주입을 요청할 수 있게 됩니다.

테스트 프레임워크에서 프레임워크의 주체는 `Jupiter`이기 때문에 생성자 주입을 한들 찾지 못하는 겁니다.

최종적으로 `Jupiter`는 생성자 매개변수를 처리할 `ParameterResolver`까지 뒤져보지만 없기 때문에 예외를 발생시킵니다.

---

두번째 질문에 대한 답변은 에러코드가 함께 첨부 되면 좋을 것 같은데, 구체적으로 어떤 상황인지를 알 수가 없네요..

> 참조자료
> > [JUnit 5 + Kotlin 테스트 클래스에서 생성자 주입 이슈](https://minkukjo.github.io/framework/2020/06/28/JUnit-23/)

**Comment by BvrPark** [comment]
2번 저 부분은 저도 처음에 오류가 났다가
```java
@TestPropertySource(locations = "classpath:application-test.yaml")
// 이 부분을

@TestPropertySource(properties = {"spring.config.location = classpath:application-test.yaml"})
// 이렇게 바꾸니까 그냥 main 폴더에서만 yaml 설정해줘도 되긴하던데 이게 해결 방법이 맞는지는 모르겠습니다
```

**Reply by Irisation23** [reply]
https://www.baeldung.com/spring-yaml-propertysource

읽어 보시면 좋을 것 같습니다.

**Comment by Irisation23** [comment]
### 금주 CheckPoint 

- 다음 참여 시 진행 예정
